src/protocol/tx-validity.md

```diff
@@ -178,6 +178,11 @@ def metadata_gas_fees(tx) -> int:
         if tx.upgradePurpose.type == UpgradePurposeType.ConsensusParameters:
             # add intrinsic cost of calculating the consensus parameters hash
             total += sha256_gas_fee(size(tx.witnesses[tx.upgradePurpose.witnessIndex].data))
+    elif tx.type == TransactionType.Upload:
+        # add intrinsic cost of calculating the root based on the number of bytecode subsections
+        total += contract_state_root_gas_fee(tx.subsectionsNumber)
```

Comment:
Is this function explained somewhere? What does it do, exactly? The MT grows `log(subsections)`, right?

Reply:
We don't have an explanation for this function. Basically, all functions that are used here are based on our benchmark.